File: tests/fixtures/ecosystems/scala_comprehensive/Functional.scala
package comprehensive

object Functional {
  // Higher-order functions
  def transform[A, B](items: List[A])(f: A => B): List[B] = items.map(f)

  def filter[A](items: List[A])(p: A => Boolean): List[A] = items.filter(p)

  def reduce[A](items: List[A])(f: (A, A) => A): A = items.reduce(f)

  // Pattern matching
  def describe(x: Any): String = x match {
    case i: Int if i > 0 => s"Positive: $i"
    case s: String => s"String: $s"
    case (a, b) => s"Tuple: ($a, $b)"
    case head :: tail => s"List starting with $head"
    case None => "Nothing"
    case Some(v) => s"Something: $v"
    case _ => "Unknown"
  }

  // For comprehension
  def combinations(xs: List[Int], ys: List[Int]): List[(Int, Int)] =
    for {
      x <- xs
      y <- ys
      if x + y > 5
    } yield (x, y)

  // Partial functions
  val safeDiv: PartialFunction[(Int, Int), Double] = {
    case (a, b) if b != 0 => a.toDouble / b
  }

  // Currying
  def multiply(a: Int)(b: Int): Int = a * b
  val double: Int => Int = multiply(2)
}

File: tests/fixtures/ecosystems/haskell_comprehensive/Functional.hs
module Functional where

import Data.List (foldl')

-- Higher-order functions
applyTwice :: (a -> a) -> a -> a
applyTwice f x = f (f x)

-- Map and filter
transform :: (a -> b) -> (b -> Bool) -> [a] -> [b]
transform f p = filter p . map f

-- Fold
sumAll :: [Int] -> Int
sumAll = foldl' (+) 0

productAll :: [Int] -> Int
productAll = foldl' (*) 1

-- Function composition
pipeline :: [Int] -> [Int]
pipeline = filter even . map (*2) . filter (>0)

-- Lambda
incrementAll :: [Int] -> [Int]
incrementAll = map (\x -> x + 1)

-- Currying
multiply :: Int -> Int -> Int
multiply a b = a * b

double :: Int -> Int
double = multiply 2

triple :: Int -> Int
triple = multiply 3
